### PR TITLE
feat(Core/Database): implement db loader

### PR DESCRIPTION
**Pull request by Winfidonarleyan**
## Changes Proposed:
- Partially port https://github.com/TrinityCore/TrinityCore/commit/2bd1c8541d638f5041a92f491f02919270981bc2

## Issues Addressed:
- Part #3739

### Ubuntu 20.04
- `authserver`

image
- `worldserver`

image
image

### Windows
- `authserver`
image

- `worldserver`
image
image

## How to Test the Changes:
1. Startup
2. Check world, maybe wait for 1-5 mins

## TODO
- [x] license

## Target Branch(es):
- [x] Master
 
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by yuanf225** [comment]
Waiting to test this PR, great function

**Comment by Winfidonarleyan** [comment]
@yuanf225 Try :D

**Comment by yuanf225** [comment]
I'm taking the day off. I'll test it tomorrow

**Comment by yuanf225** [comment]
[Compile error.txt](https://github.com/azerothcore/azerothcore-wotlk/files/5941026/Compile.error.txt)
Test environment: Ubuntu18.04
Install all AC dependencies：
./acore.sh install-deps
Build everything from scratch：
./acore.sh compiler all

The attachment is the error message of the compilation process. I am not sure whether it is this PR problem or other problems. I will try to pull a clean core under the same environment for compilation attempt

**Comment by Winfidonarleyan** [comment]
@yuanf225 need clean core and check github actions

**Comment by Winfidonarleyan** [comment]
Test this in `Ubuntu 20.04` and `Windows 10`

File: src/common/Database/DatabaseEnv.cpp [hunk 1]
```diff
@@ -0,0 +1,22 @@
+/*
+ * This file is part of the WarheadCore Project. See AUTHORS file for Copyright information
```
Comment (FrancescoBorzi) [inline src/common/Database/DatabaseEnv.cpp:2]: please update this. It's ok to mention Warhead but make this standard as the other files

File: src/common/Database/DatabaseLoader.cpp [hunk 2]
```diff
@@ -0,0 +1,158 @@
+/*
+ * This file is part of the WarheadCore Project. See AUTHORS file for Copyright information
```
Comment (FrancescoBorzi) [inline src/common/Database/DatabaseLoader.cpp:2]: same

File: src/common/Database/DatabaseLoader.h [hunk 3]
```diff
@@ -0,0 +1,68 @@
+/*
+ * This file is part of the WarheadCore Project. See AUTHORS file for Copyright information
```
Comment (FrancescoBorzi) [inline src/common/Database/DatabaseLoader.h:2]: same

**Comment by FrancescoBorzi** [comment]
@azerothcore/testers @pklloveyou @mpfans please test this

**Comment by claudiodfc** [comment]
I didn't have much success compiling on Windows, Visual Studio Comunity 19:

```
Build started...
1>------ Build started: Project: revision.h, Configuration: Debug x64 ------
2>------ Skipped Build: Project: clean_cotire, Configuration: Debug x64 ------
2>Project not selected to build for this solution configuration 
3>------ Build started: Project: authserver, Configuration: Debug x64 ------
4>------ Build started: Project: worldserver, Configuration: Debug x64 ------
3>Main.obj : error LNK2019: unresolved external symbol "public: bool __cdecl DatabaseLoader::Load(void)" (?Load@DatabaseLoader@@QEAA_NXZ) referenced in function "bool __cdecl StartDB(void)" (?StartDB@@YA_NXZ)

3>Main.obj : error LNK2019: unresolved external symbol "public: class DatabaseLoader & __cdecl DatabaseLoader::AddDatabase<class LoginDatabaseConnection>(class DatabaseWorkerPool<class LoginDatabaseConnection> &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (??$AddDatabase@VLoginDatabaseConnection@@@DatabaseLoader@@QEAAAEAV0@AEAV?$DatabaseWorkerPool@VLoginDatabaseConnection@@@@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z) referenced in function "bool __cdecl StartDB(void)" (?StartDB@@YA_NXZ)

3>Main.obj : error LNK2001: unresolved external symbol "class DatabaseWorkerPool<class LoginDatabaseConnection> LoginDatabase" (?LoginDatabase@@3V?$DatabaseWorkerPool@VLoginDatabaseConnection@@@@A)

3>AuthSocket.obj : error LNK2001: unresolved external symbol "class DatabaseWorkerPool<class LoginDatabaseConnection> LoginDatabase" (?LoginDatabase@@3V?$DatabaseWorkerPool@VLoginDatabaseConnection@@@@A)

3>shared.lib(RealmList.obj) : error LNK2001: unresolved external symbol "class DatabaseWorkerPool<class LoginDatabaseConnection> LoginDatabase" (?LoginDatabase@@3V?$DatabaseWorkerPool@VLoginDatabaseConnection@@@@A)

3>common.lib(Log.obj) : error LNK2001: unresolved external symbol "class DatabaseWorkerPool<class LoginDatabaseConnection> LoginDatabase" (?LoginDatabase@@3V?$DatabaseWorkerPool@VLoginDatabaseConnection@@@@A)

3>D:\_azerothcore\_acore_build\bin\Debug\authserver.exe : fatal error LNK1120: 3 unresolved externals

3>Done building project "authserver.vcxproj" -- FAILED.

4>Master.obj : error LNK2019: unresolved external symbol "public: bool __cdecl DatabaseLoader::Load(void)" (?Load@DatabaseLoader@@QEAA_NXZ) referenced in function "private: bool __cdecl Master::_StartDB(void)" (?_StartDB@Master@@AEAA_NXZ)

4>Master.obj : error LNK2019: unresolved external symbol "public: class DatabaseLoader & __cdecl DatabaseLoader::AddDatabase<class LoginDatabaseConnection>(class DatabaseWorkerPool<class LoginDatabaseConnection> &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (??$AddDatabase@VLoginDatabaseConnection@@@DatabaseLoader@@QEAAAEAV0@AEAV?$DatabaseWorkerPool@VLoginDatabaseConnection@@@@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z) referenced in function "private: bool __cdecl Master::_StartDB(void)" (?_StartDB@Master@@AEAA_NXZ)

4>Master.obj : error LNK2019: unresolved external symbol "public: class DatabaseLoader & __cdecl DatabaseLoader::AddDatabase<class CharacterDatabaseConnection>(class DatabaseWorkerPool<class CharacterDatabaseConnection> &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (??$AddDatabase@VCharacterDatabaseConnection@@@DatabaseLoader@@QEAAAEAV0@AEAV?$DatabaseWorkerPool@VCharacterDatabaseConnection@@@@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z) referenced in function "private: bool __cdecl Master::_StartDB(void)" (?_StartDB@Master@@AEAA_NXZ)

4>Master.obj : error LNK2019: unresolved external symbol "public: class DatabaseLoader & __cdecl DatabaseLoader::AddDatabase<class WorldDatabaseConnection>(class DatabaseWorkerPool<class WorldDatabaseConnection> &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (??$AddDatabase@VWorldDatabaseConnection@@@DatabaseLoader@@QEAAAEAV0@AEAV?$DatabaseWorkerPool@VWorldDatabaseConnection@@@@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z) referenced in function "private: bool __cdecl Master::_StartDB(void)" (?_StartDB@Master@@AEAA_NXZ)

4>D:\_azerothcore\_acore_build\bin\Debug\worldserver.exe : fatal error LNK1120: 4 unresolved externals

4>Done building project "worldserver.vcxproj" -- FAILED.

5>------ Skipped Build: Project: INSTALL, Configuration: Debug x64 ------

5>Project not selected to build for this solution configuration 
========== Build: 1 succeeded, 2 failed, 14 up-to-date, 2 skipped ==========```

**Comment by Winfidonarleyan** [comment]
> I didn't have much success compiling on Windows, Visual Studio Comunity 19:
> 
> ```
> Build started...
> 1>------ Build started: Project: revision.h, Configuration: Debug x64 ------
> 2>------ Skipped Build: Project: clean_cotire, Configuration: Debug x64 ------
> 2>Project not selected to build for this solution configuration 
> 3>------ Build started: Project: authserver, Configuration: Debug x64 ------
> 4>------ Build started: Project: worldserver, Configuration: Debug x64 ------
> 3>Main.obj : error LNK2019: unresolved external symbol "public: bool __cdecl DatabaseLoader::Load(void)" (?Load@DatabaseLoader@@QEAA_NXZ) referenced in function "bool __cdecl StartDB(void)" (?StartDB@@YA_NXZ)
> 
> 3>Main.obj : error LNK2019: unresolved external symbol "public: class DatabaseLoader & __cdecl DatabaseLoader::AddDatabase<class LoginDatabaseConnection>(class DatabaseWorkerPool<class LoginDatabaseConnection> &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (??$AddDatabase@VLoginDatabaseConnection@@@DatabaseLoader@@QEAAAEAV0@AEAV?$DatabaseWorkerPool@VLoginDatabaseConnection@@@@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z) referenced in function "bool __cdecl StartDB(void)" (?StartDB@@YA_NXZ)
> 
> 3>Main.obj : error LNK2001: unresolved external symbol "class DatabaseWorkerPool<class LoginDatabaseConnection> LoginDatabase" (?LoginDatabase@@3V?$DatabaseWorkerPool@VLoginDatabaseConnection@@@@A)
> 
> 3>AuthSocket.obj : error LNK2001: unresolved external symbol "class DatabaseWorkerPool<class LoginDatabaseConnection> LoginDatabase" (?LoginDatabase@@3V?$DatabaseWorkerPool@VLoginDatabaseConnection@@@@A)
> 
> 3>shared.lib(RealmList.obj) : error LNK2001: unresolved external symbol "class DatabaseWorkerPool<class LoginDatabaseConnection> LoginDatabase" (?LoginDatabase@@3V?$DatabaseWorkerPool@VLoginDatabaseConnection@@@@A)
> 
> 3>common.lib(Log.obj) : error LNK2001: unresolved external symbol "class DatabaseWorkerPool<class LoginDatabaseConnection> LoginDatabase" (?LoginDatabase@@3V?$DatabaseWorkerPool@VLoginDatabaseConnection@@@@A)
> 
> 3>D:\_azerothcore\_acore_build\bin\Debug\authserver.exe : fatal error LNK1120: 3 unresolved externals
> 
> 3>Done building project "authserver.vcxproj" -- FAILED.
> 
> 4>Master.obj : error LNK2019: unresolved external symbol "public: bool __cdecl DatabaseLoader::Load(void)" (?Load@DatabaseLoader@@QEAA_NXZ) referenced in function "private: bool __cdecl Master::_StartDB(void)" (?_StartDB@Master@@AEAA_NXZ)
> 
> 4>Master.obj : error LNK2019: unresolved external symbol "public: class DatabaseLoader & __cdecl DatabaseLoader::AddDatabase<class LoginDatabaseConnection>(class DatabaseWorkerPool<class LoginDatabaseConnection> &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (??$AddDatabase@VLoginDatabaseConnection@@@DatabaseLoader@@QEAAAEAV0@AEAV?$DatabaseWorkerPool@VLoginDatabaseConnection@@@@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z) referenced in function "private: bool __cdecl Master::_StartDB(void)" (?_StartDB@Master@@AEAA_NXZ)
> 
> 4>Master.obj : error LNK2019: unresolved external symbol "public: class DatabaseLoader & __cdecl DatabaseLoader::AddDatabase<class CharacterDatabaseConnection>(class DatabaseWorkerPool<class CharacterDatabaseConnection> &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (??$AddDatabase@VCharacterDatabaseConnection@@@DatabaseLoader@@QEAAAEAV0@AEAV?$DatabaseWorkerPool@VCharacterDatabaseConnection@@@@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z) referenced in function "private: bool __cdecl Master::_StartDB(void)" (?_StartDB@Master@@AEAA_NXZ)
> 
> 4>Master.obj : error LNK2019: unresolved external symbol "public: class DatabaseLoader & __cdecl DatabaseLoader::AddDatabase<class WorldDatabaseConnection>(class DatabaseWorkerPool<class WorldDatabaseConnection> &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (??$AddDatabase@VWorldDatabaseConnection@@@DatabaseLoader@@QEAAAEAV0@AEAV?$DatabaseWorkerPool@VWorldDatabaseConnection@@@@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z) referenced in function "private: bool __cdecl Master::_StartDB(void)" (?_StartDB@Master@@AEAA_NXZ)
> 
> 4>D:\_azerothcore\_acore_build\bin\Debug\worldserver.exe : fatal error LNK1120: 4 unresolved externals
> 
> 4>Done building project "worldserver.vcxproj" -- FAILED.
> 
> 5>------ Skipped Build: Project: INSTALL, Configuration: Debug x64 ------
> 
> 5>Project not selected to build for this solution configuration 
> ========== Build: 1 succeeded, 2 failed, 14 up-to-date, 2 skipped ==========```
> ```

https://github.com/azerothcore/azerothcore-wotlk/pull/4431/checks?check_run_id=2140280326

**Comment by Winfidonarleyan** [comment]
@claudiodfc try clean build

**Comment by Yehonal** [comment]
Does it support the current sql format with the "procedures"? Would be great if db_assembler can be still supported.

In any case, we should keep our flow of using a "pending sql" that will be automatically moved/renamed to a proper filename format. This approach avoided any kind of collision between our sql dumps and it's better to not remove that one at least. I don't see any changes to that system here, which is good, but it's good to keep in mind it.

**Comment by claudiodfc** [comment]
@Winfidonarleyan The build is passing on MVS2016 but I'm compiling with 2019 where the linker fails here with this error

https://docs.microsoft.com/en-us/cpp/error-messages/tool-errors/linker-tools-error-lnk2019?view=msvc-160

**Comment by FrancescoBorzi** [comment]
> Does it support the current sql format with the "procedures"? Would be great if db_assembler can be still supported.
> 
> In any case, we should keep our flow of using a "pending sql" that will be automatically moved/renamed to a proper filename format. This approach avoided any kind of collision between our sql dumps and it's better to not remove that one at least. I don't see any changes to that system here, which is good, but it's good to keep in mind it.

I don't think the db_assembler  is affected by this. If it is, it shouldn't. Also, the CI is passing fine.

Or maybe I misunderstood what you mean?

**Comment by claudiodfc** [comment]
Visual studio 2016 is different from 2019 which is what Azerothcore recommends. Therefore, this PR fails on 2019 but not on 2016

**Comment by Winfidonarleyan** [comment]
@claudiodfc https://gist.github.com/Winfidonarleyan/b5257e2b8b629c09dadb0395306d892d

**Comment by Winfidonarleyan** [comment]
And, about `auto update db`
- This PR rework ONLY load databalse
- NO! sql applying and etc

**Comment by claudiodfc** [comment]
@Winfidonarleyan Can you fix the conflicting files, please? I'll test when done 👍🏼

**Comment by Winfidonarleyan** [comment]
Today

**Comment by Winfidonarleyan** [comment]
@claudiodfc solved conflicts, ready to test

**Comment by yuanf225** [comment]
> I can't compile this man, I re-instaled VSC2019 and the same errors are happening to me, clean project and I can't compile this

What parameters do you use for compilation

**Comment by claudiodfc** [comment]
I'm doing it on release x64

**Comment by yuanf225** [comment]
I will also try to compile at night

**Comment by Winfidonarleyan** [comment]
Now I got the same error, for eazy fix
1. Clean CMake cache
image
2. Configure
3. Build

```
10>------ Сборка начата: проект: authserver, Конфигурация: RelWithDebInfo x64 ------
10>Generating CXX prefix header F:\Git\azerothcore-wotlk\build\src\server\authserver\RelWithDebInfo\cotire\authserver_CXX_prefix.hxx
10>Building Custom Rule F:/Git/azerothcore-wotlk/src/server/authserver/CMakeLists.txt
10>Main.cpp
10>AuthCodes.cpp
10>AuthSocket.cpp
10>TOTP.cpp
10>WheatyExceptionReport.cpp
10>authserver.vcxproj -> F:\Git\azerothcore-wotlk\build\bin\RelWithDebInfo\authserver.exe
========== Сборка: успешно: 10, с ошибками: 0, без изменений: 0, пропущено: 0 ==========
```

**Comment by Nefertumm** [comment]
Great job as usual Winfi. I'm really looking forward to see more of this!

**Comment by Winfidonarleyan** [comment]
@claudiodfc I installed the current version for today

**Comment by Kitzunu** [comment]
Worldserver starts as normal

**Comment by Winfidonarleyan** [comment]
Ready to review

File: src/common/Database/DatabaseWorkerPool.cpp [hunk 4]
```diff
@@ -99,6 +95,82 @@ void DatabaseWorkerPool<T>::Close()
...
+    for (uint8 i = 0; i < numConnections; ++i)
+    {
+        T* t;
+
+        if (type == IDX_ASYNC)
+            t = new T(_queue, *_connectionInfo);
+        else if (type == IDX_SYNCH)
+            t = new T(*_connectionInfo);
+        else
+            ABORT();
```
Comment (FrancescoBorzi) [inline src/common/Database/DatabaseWorkerPool.cpp:111]: reason to abort? also include brackets please

Comment (FrancescoBorzi) [reply src/common/Database/DatabaseWorkerPool.cpp:111]: `ABORT("some reason");`

File: src/common/Database/DatabaseWorkerPool.cpp [hunk 5]
```diff
@@ -99,6 +95,82 @@ void DatabaseWorkerPool<T>::Close()
...
+
+        _connections[type][i] = t;
+        ++_connectionCount[type];
+
+        uint32 error = t->Open();
+
+        if (!error)
+        {
+            if (mysql_get_server_version(t->GetHandle()) < MIN_MYSQL_SERVER_VERSION)
```
Comment (FrancescoBorzi) [inline src/common/Database/DatabaseWorkerPool.cpp:120]: please make sure it works with all versions of MariaDB, including the ones based on MySQL 5.5

File: src/common/Database/DatabaseLoader.cpp [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+                {
+                    error = pool.Open();
+
+                    if (error == CR_CONNECTION_ERROR)
+                    {
+                        sleepThread();
+                        count++;
+                    }
+                    else
+                        break;
```
Comment (FrancescoBorzi) [inline src/common/Database/DatabaseLoader.cpp:65]: brackets

File: src/common/Database/DatabaseLoader.cpp [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+bool DatabaseLoader::Load()
+{
+    if (!OpenDatabases())
+        return false;
+
+    if (!PrepareStatements())
+        return false;
+
+    return true;
```
Comment (FrancescoBorzi) [inline src/common/Database/DatabaseLoader.cpp:109]: isn't this the same as:

```suggestion
    return OpenDatabases() && PrepareStatements();
```

File: src/common/Database/DatabaseLoader.h [hunk 8]
```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (C) 2016+ AzerothCore <www.azerothcore.org>, released under GNU GPL v2 license, you may redistribute it and/or modify it under version 2 of the License, or (at your option), any later version.
```
Comment (FrancescoBorzi) [inline src/common/Database/DatabaseLoader.h:2]: new file can be AGPL v3

File: src/common/Database/DatabaseLoader.cpp [hunk 9]
```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (C) 2016+ AzerothCore <www.azerothcore.org>, released under GNU GPL v2 license, you may redistribute it and/or modify it under version 2 of the License, or (at your option), any later version.
```
Comment (FrancescoBorzi) [inline src/common/Database/DatabaseLoader.cpp:2]: new file can be AGPL v3

File: src/common/Database/DatabaseEnv.cpp [hunk 10]
```diff
@@ -0,0 +1,10 @@
+/*
+ * Copyright (C) 2016+ AzerothCore <www.azerothcore.org>, released under GNU GPL v2 license, you may redistribute it and/or modify it under version 2 of the License, or (at your option), any later version.
```
Comment (FrancescoBorzi) [inline src/common/Database/DatabaseEnv.cpp:2]: new file can be AGPL v3

File: src/common/Database/DatabaseWorkerPool.cpp [hunk 11]
```diff
@@ -99,6 +95,82 @@ void DatabaseWorkerPool<T>::Close()
...
+            t->LockIfReady();
+
+            if (!t->PrepareStatements())
+            {
+                t->Unlock();
+                Close();
+                return false;
+            }
+            else
+                t->Unlock();
```
Comment (FrancescoBorzi) [inline src/common/Database/DatabaseWorkerPool.cpp:161]: brackets

File: src/common/Database/MySQLConnection.cpp [hunk 12]
```diff
@@ -106,59 +107,42 @@ bool MySQLConnection::Open()
...
-                sLog->outSQLDriver("MySQL server ver: %s ", mysql_get_server_info(m_Mysql));
-                // MySQL version above 5.1 IS required in both client and server and there is no known issue with different versions above 5.1
-                // if (mysql_get_server_version(m_Mysql) != mysql_get_client_version())
-                //     sLog->outInfo(LOG_FILTER_SQL, "[WARNING] MySQL client/server version mismatch; may conflict with behaviour of prepared statements.");
-            }
+            sLog->outSQLDriver("MySQL client library: %s", mysql_get_client_info());
+            sLog->outSQLDriver("MySQL server ver: %s ", mysql_get_server_info(m_Mysql));
+
+            if (mysql_get_server_version(m_Mysql) != mysql_get_client_version())
+                sLog->outSQLDriver("[WARNING] MySQL client/server version mismatch; may conflict with behaviour of prepared statements.");
```
Comment (FrancescoBorzi) [inline src/common/Database/MySQLConnection.cpp:128]: brackets

File: src/common/Database/DatabaseWorkerPool.cpp (hunk 5 above)
Comment (Kitzunu) [reply src/common/Database/DatabaseWorkerPool.cpp:120]: Mysql has dropped support for 5.5 and 5.6, and so have we long ago.

File: src/common/Database/DatabaseLoader.cpp (hunk 7 above)
Comment (Winfidonarleyan) [reply src/common/Database/DatabaseLoader.cpp:109]: very good solution

File: src/common/Database/DatabaseWorkerPool.cpp (hunk 5 above)
Comment (Winfidonarleyan) [reply src/common/Database/DatabaseWorkerPool.cpp:120]: Work with MariaDB 10.3 - 10.5

File: src/common/Database/DatabaseWorkerPool.cpp (hunk 4 above)
Comment (Winfidonarleyan) [reply src/common/Database/DatabaseWorkerPool.cpp]: `ABORT_MSG` - not implement in core now

**Review by FrancescoBorzi — APPROVED** [review]

**Pull request merged**